src/tracer.cc

```diff
@@ -369,30 +371,225 @@ TraceManager::Trace::InitTracer(
+      // and can not be a parent of any sub-span. If parent_id=0,
+      // then current model is either a standalone model, or an ensemble model.
+      // In both of the above cases, the parent of the new request sub-span
+      // is the kRootSpan.
+      std::string parent_span_key = kRootSpan;
+      if (parent_id != 0) {
+        // If parent_id > 0, then this is a child trace, spawned from
+        // the ensamble's main request. For this instance, the parent
+        // span is the ensembles's request span.
+        parent_span_key = kRequestSpan + std::to_string(parent_id);
+      }
+      auto request_span = this->InitSpan(
+          model_name,
+          this->time_offset_ + std::chrono::nanoseconds{timestamp_ns},
+          timestamp_ns, false /*is_root_span*/, parent_span_key);
+      request_span->SetAttribute("triton.model_name", model_name);
+      request_span->SetAttribute("triton.model_version", model_version);
+      request_span->SetAttribute("triton.trace_id", id);
+      request_span->SetAttribute("triton.trace_parent_id", parent_id);
+      request_span->AddEvent(
+          TRITONSERVER_InferenceTraceActivityString(activity),
+          this->time_offset_ + std::chrono::nanoseconds{timestamp_ns});
+      // Adding span to the OpenTelemetry context, to keep it alive and active
+      // until we end it
+      this->otel_context_ =
+          this->otel_context_.SetValue(current_span_key, request_span);
+      break;
+    }
+    case TRITONSERVER_TRACE_COMPUTE_START: {
+      // Any compute span is a child of a request span with the same id.
+      auto compute_span = this->InitSpan(
+          "compute",
+          this->time_offset_ + std::chrono::nanoseconds{timestamp_ns},
+          timestamp_ns, false /*is_root_span*/,
+          kRequestSpan + std::to_string(id));
+      compute_span->AddEvent(
+          TRITONSERVER_InferenceTraceActivityString(activity),
+          this->time_offset_ + std::chrono::nanoseconds{timestamp_ns});
+      // Adding span to the OpenTelemetry context, to keep it alive and active
+      // until we end it
+      this->otel_context_ =
+          this->otel_context_.SetValue(current_span_key, compute_span);
+      break;
+    }
+    case TRITONSERVER_TRACE_QUEUE_START:
+    case TRITONSERVER_TRACE_COMPUTE_INPUT_END:
+    case TRITONSERVER_TRACE_COMPUTE_OUTPUT_START: {
+      if (this->otel_context_.HasKey(current_span_key)) {
+        auto span = opentelemetry::nostd::get<
+            opentelemetry::nostd::shared_ptr<otel_trace_api::Span>>(
+            this->otel_context_.GetValue(current_span_key));
+        span->AddEvent(
+            TRITONSERVER_InferenceTraceActivityString(activity),
+            this->time_offset_ + std::chrono::nanoseconds{timestamp_ns});
+      }
+      break;
+    }
+    case TRITONSERVER_TRACE_COMPUTE_END:
+    case TRITONSERVER_TRACE_REQUEST_END: {
+      if (this->otel_context_.HasKey(current_span_key)) {
+        auto span = opentelemetry::nostd::get<
```

Comment:
L537-L543 are identical to L525-L531. 
Can you create a new function for AddEvent? 
You can include an option in this function to whether or not to call EndSpan.

Reply:
I refactored the code.